# Introduction to Python

Throughout this entire notebook you should be experimenting with the code in the non-text cells. A great way to begin to get a feel for Python is by playing with it. So have some fun by changing the values in the cells and then running them again with Shift-Enter. 

At the end of each section there will be some questions to help further your understanding. Remember, in Python we can always manually test things by trying them out; however, you should try to think about the answers to these questions before you run some code. This way you can check and verify your understanding of the section's topic.

### Numeric Variable Types

The first thing you are going to learn about are base numeric variable types in Python. There are a number of these base numeric variable types built into Python, and all of them represent a very simple idea, numbers. The ones that we will be looking at today are `ints`, short for integers, `floats`, short for floating point decimal numbers, and `complex`, which contain real and imaginary parts stored as floats.

In [1]:
7 # Int

7

In [2]:
7.5 # Float

7.5

In [3]:
complex(3, 4) # complex

(3+4j)

### Types

An important characteristic of Python is that it is a duck typed language. What does this mean? The name duck comes from the classic "If it walks like a duck, and quacks like a duck, then it must be a duck" adage. As applied to our situation, it simply means that Python will determine what it thinks is the best type to call a variable when you use it, unless explicitly told otherwise.

To inspect what type Python thinks a numeric (or anything else is), you can pass it to the `type()` function. Let's see what we get out when we pass numbers of various types to this function.

In [4]:
type(7)

int

In [5]:
type(7.5)

float

In [6]:
type(complex(3, 4))

complex

As you can see, Python assumes that a number with no decimal point is an `int`, those with a decimal point a `float`, and (surprise!) those from the `complex()` constructor as `complex`.

Frequently, these subtle differences wont matter too much. However, there will be times when this implementation detail will make you think that something will work, when really it won't. Knowing how to check the type of something will help you solve any of these potential problems.

**Type Questions**

What are the types of the following numerics?
1. `10`
2. `100.32`
3. `complex(-5, -10.2)`

### Numeric Operations

At its base level, Python is really just an awesome calculator that can do way more stuff than addition and subtraction. But, let's focus on that functionality for now.

All of the simple operations that you think should be available are available. Addition, subtraction, multiplication, division and exponentiation are all accessible via `+`, `-`, `*`, `/` (and `//`), and `**`, respectively.

In [7]:
7 + 8

15

In [8]:
7 - 8

-1

In [9]:
7 * 8

56

In [11]:
7 / 8 # floating point division

0.875

In [10]:
7 // 8 # floor/integer division

0

In [11]:
7 ** 8

5764801

All of these operations output exactly what we think they would, except two

The fourth cell returned a float even though we passed it two integers.  From Python 3 on (make note of this in case you ever use Python 2), `/` will return a float even if two integers are passed.

The fifth cell returned an int of `0` even though we know that it should be 0.875.  This is because the `//` is for floor division rather than floating point division.  All it does is perform division and truncate the result. So, where `7 / 8.` gave us `0.875`, `7 // 8.` cuts off after the `0.` giving us `0.0`.

Let's visit all aspects of this problem. Make sure to try out different combinations of the following situations.

In [2]:
7 // 8 # Int / Int

0

In [14]:
7.0 // 8.0 # Float / Float

0.0

In [3]:
7 / 8.0 # Int / Float

0.875

In [4]:
7 / 8. # Int / Float

0.875

In [15]:
int(7 / 8.) # Int(Int / Float)

0

In [1]:
7 // 8 # Int // Int

0

Notice that once we made one of the numbers in the operation a `float`, Python realized that it should return a `float` from the operation, and we got 0.875 (as expected). This exact same procedure was followed the third time, but the trailing 0 on the `float` version of 8 was left off. Python doesn't need to see anything following the decimal point to know it should be interpreting the number as a `float`, so long as the "." is there. 

Similarly, we can manually cast the output of the 5th operation as an `int` by passing the result directly to the `int()` constructor. Manually casting in this way can be very useful when Python is interpreting things differently than you would like it to. 

The last operation that we will go over is the modular division operator, `%`. This operation is the sibling to `//`. Consider the example:

In [6]:
71 // 7

10

We see that the `/` gives us the integer number of times that 7 goes into 71. But, there is still a remainder of one. The way we get the remainder of integer division is with the `%` operator.

In [7]:
71 % 7

1

Finally, note that you can change the order in which operations get performed by using parenthesis `()`, just as you can in algebra.

In [8]:
4 + 5 * 3

19

In [9]:
(4 + 5) * 3

27

**Operation Questions**

What do you think the results of the following computations will be?
1. `8.0 - 7`
2. `8 * 0.25`
3. `5 ** 2`
4. `(4 + 3) / 6`
5. `(20 / 3) % 2`

### Variables

One of the most powerful constructs in programming is the ability to store arbitrary values in what we call variables. You can think of variable assignment as giving a name to something so that it can be accessed later by different parts of your program.

In Python, variable assignment occurs with the `=` operator. To assign a value to a variable name (i.e. declare it), you simply put the variable name on the left side of the `=` and the value you want to associate with that variable name on the right side. Once this has happened, you can access the value in the variable simply by using it's name somewhere later in your code or IPython session.

In [21]:
x = 5
y = 2

In [22]:
x

5

In [23]:
x - 4

1

In [24]:
x + y

7

The name you can give a variable can technically be any contiguous set of characters, but there are some conventions followed in Python and programming in general. Python follows a variable naming convention called snake case. To write something in snake case, simply use a `_` anywhere you would use a space, and make sure every word is lower case. For example, `this_is_a_variable`. Giving variables good names makes your code more readable and therefore maintainable. There is a big difference between seeing a variable called `degrees` and one called `y`. You should strive to give your variables well-defined, succinct names.

There are of course cases where using less than descriptive variable names follows convention and are, therefore, just fine to use. A common example is the use of `i` to keep track of an index (we will cover this idea next week). Because of its prevalent usage for indexing, it is usually easy to understand what is happening in that context when all you see is the varible name `i`. Here, the lack of descriptiveness is okay. The important thing is that the code is **understandable**.

Note that we saw no output from either `x = 5` or `y = 2` above. This is because the return value that would have been printed as output was assigned to the variables `x` and `y`, respectively. This is why we had to view them in the next lines.

A large part of variables' power is the fact that they can change (vary, if you will). This allows us to use a single variable name to keep track of a specific thing throughout the life of a program. Remember how we assigned the value `5` to `x` above? The exact same syntax can be used to change the value stored in the variable.

Say we want to make the value of `x` five more than it currently is. All we need to do is have x be assigned the value that results from adding 5 to `x`.

In [25]:
x = x + 5
x

10

Notice how the first line above is formatted. Python knows that the `=` means variable assignment, so when it sees the first line it evaluates the right side of the equals and then puts that value in `x`, even though `x` is part of the calculation on the right side. `x` is now connected with this new value and the old value is gone.

Changing variables in this way occurs so commonly that there is built-in shorthand for it. The result of the first line could have been achieved with `x += 5`. This *syntactic sugar* is available for all the simple operations `+`, `-`, `*`, `/`, `**`, and `%` that we covered earlier.

**Variable Questions**

Consider the following code:

```
x = 5
y = 8
x += y
y = x - 3
x -= y - 5
```

What are the values of x and y after each line?